File: library/dialog/src/main/java/com/github/lany192/dialog/SimpleDialog.java
package com.github.lany192.dialog;

import android.text.TextUtils;
import android.text.method.MovementMethod;
import android.text.method.ScrollingMovementMethod;
import android.view.Gravity;
import android.view.View;

import androidx.annotation.ColorRes;
import androidx.annotation.StringRes;

import com.github.lany192.dialog.databinding.DialogSimpleBinding;
import com.github.lany192.interfaces.OnSimpleListener;

import java.lang.reflect.Modifier;

//    SimpleDialog dialog = new SimpleDialog();
//    dialog.setTitle("提示");
//    dialog.setMessage("猜猜我是谁");
//    dialog.setRightBtn("确定", new OnSimpleListener() {
//        @Override
//        public void onCallback() {
//
//        }
//    });
//    dialog.setLeftBtn("取消", new OnSimpleListener() {
//        @Override
//        public void onCallback() {
//
//        }
//    });
//    dialog.show();

public class SimpleDialog extends BaseDialog<DialogSimpleBinding> {
    private OnSimpleListener mOnRightListener;
    private OnSimpleListener mOnLeftListener;
    private CharSequence mTitle;
    private CharSequence mMessage;
    private CharSequence mLeftText;
    private CharSequence mRightText;
    private boolean isShowDivider;
    @ColorRes
    private int mTitleColor = R.color.text_1level;
    private float titleSize = 16;
    private float mMsgTextSize = 14;
    @ColorRes
    private int mRightTextColor = 0;
    @ColorRes
    private int mLeftTextColor = 0;

    private MovementMethod movementMethod;

    @Override
    protected void init() {
        if (TextUtils.isEmpty(mTitle)) {
            binding.title.setVisibility(View.GONE);
        } else {
            binding.title.setText(mTitle);
            binding.title.setTextSize(titleSize);
            binding.title.setVisibility(View.VISIBLE);
            binding.title.setTextColorId(mTitleColor);
        }
        if (!TextUtils.isEmpty(mMessage)) {
            binding.content.setText(mMessage);
            binding.content.post(() -> {
                if (binding.content.getLineCount() > 1) {
                    binding.content.setGravity(Gravity.LEFT);
                } else {
                    binding.content.setGravity(Gravity.CENTER);
                }
            });
            binding.content.setTextSize(mMsgTextSize);
            binding.content.setTextColorId(R.color.text_2level);
            if (movementMethod != null) {
                binding.content.setMovementMethod(movementMethod);
            } else {
                binding.content.setMovementMethod(ScrollingMovementMethod.getInstance());
            }
        }
        if (isShowDivider) {
            binding.divider.setVisibility(View.VISIBLE);
        } else {
            binding.divider.setVisibility(View.GONE);
        }
        if (!TextUtils.isEmpty(mRightText)) {
            binding.rightButton.setText(mRightText);
            binding.rightButton.setVisibility(View.VISIBLE);
            if (mRightTextColor != 0) {
                binding.rightButton.setTextColor(getColor(mRightTextColor));
            }
            binding.rightButton.setOnClickListener(v -> {
                cancel();
                if (null != mOnRightListener) {
                    mOnRightListener.onCallback();
                }
            });
        } else {
            binding.rightButton.setVisibility(View.GONE);
        }
        if (!TextUtils.isEmpty(mLeftText)) {
            binding.leftButton.setText(mLeftText);
            binding.leftButton.setVisibility(View.VISIBLE);
            if (mLeftTextColor != 0) {
                binding.leftButton.setTextColor(getColor(mLeftTextColor));
            }
            binding.leftButton.setOnClickListener(v -> {
                cancel();
                if (null != mOnLeftListener) {
                    mOnLeftListener.onCallback();
                }
            });
        } else {
            binding.leftButton.setVisibility(View.GONE);
        }
    }

    public void setMovementMethod(MovementMethod movementMethod) {
        this.movementMethod = movementMethod;
    }

    public void setTitle(CharSequence title) {
        this.mTitle = title;
    }

    public void setTitleColor(@ColorRes int colorIntId) {
        this.mTitleColor = colorIntId;
    }

    public void setTitleSize(float titleSize) {
        this.titleSize = titleSize;
    }

    public void setMessage(CharSequence message) {
        this.mMessage = message;
    }

    public void setMessageTextSize(float size) {
        this.mMsgTextSize = size;
    }

    public void setRightTextColor(@ColorRes int colorIntId) {
        this.mRightTextColor = colorIntId;
    }

    public void setLeftTextColor(@ColorRes int leftTextColor) {
        this.mLeftTextColor = leftTextColor;
    }

    public void setRightButton(@StringRes int rightStringRes, OnSimpleListener listener) {
        setRightButton(getString(rightStringRes), listener);
    }

    public void setLeftButton(@StringRes int leftStringRes, OnSimpleListener listener) {
        setLeftButton(getString(leftStringRes), listener);
    }

    public void setRightButton(CharSequence rightText, OnSimpleListener listener) {
        this.mRightText = rightText;
        this.mOnRightListener = listener;
    }

    public void setLeftButton(CharSequence leftText, OnSimpleListener listener) {
        this.mLeftText = leftText;
        this.mOnLeftListener = listener;
        this.isShowDivider = true;
    }

    public void setRightButton(CharSequence rightText) {
        this.mRightText = rightText;
    }

    public void setLeftButton(CharSequence leftText) {
        this.mLeftText = leftText;
        this.isShowDivider = true;
    }

    /**
     * 开启单例模式
     */
    public void setSingle() {
        setId(getUniqueDialogId());
    }

    /**
     * 根据调用类名和执行代码位置获取一个唯一的id
     */
    private long getUniqueDialogId() {
        long dialogId = System.currentTimeMillis();
        StackTraceElement[] stackTraces = (new Throwable()).getStackTrace();
        for (StackTraceElement stackTrace : stackTraces) {
            int lineNumber = stackTrace.getLineNumber();
            if (lineNumber > 0) {
                String className = stackTrace.getClassName();
                try {
                    Class<?> clazz = Class.forName(className);
                    if (!filterStackTraceClass(clazz)) {
                        int hashCode = (stackTrace.getFileName() + ":" + lineNumber).hashCode();
                        dialogId = Math.abs(hashCode);
                        log.i("对话框调用信息:当前class:" + className + ",调用代码行数：" + lineNumber);
                        break;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
        log.i("生成对话框的id:" + dialogId);
        return dialogId;
    }

    private boolean filterStackTraceClass(Class<?> clazz) {
        if (SimpleDialog.class.equals(clazz)) {
            return true;
        } else if (clazz.isInterface()) {
            return true;
        } else {
            return Modifier.isAbstract(clazz.getModifiers());
        }
    }
}

File: library/dialog/src/main/java/com/github/lany192/dialog/SimpleDialog.kt
package com.github.lany192.dialog

import android.text.TextUtils
import android.text.method.MovementMethod
import android.text.method.ScrollingMovementMethod
import android.view.Gravity
import android.view.View
import androidx.annotation.ColorRes
import com.github.lany192.dialog.databinding.DialogSimpleBinding
import com.github.lany192.interfaces.OnSimpleListener
import java.lang.reflect.Modifier
import kotlin.math.abs

// java ----------------------------------------
//    SimpleDialog dialog = new SimpleDialog();
//    dialog.setTitle("提示");
//    dialog.setMessage("猜猜我是谁");
//    dialog.setRightBtn("确定", new OnSimpleListener() {
//        @Override
//        public void onCallback() {
//
//        }
//    });
//    dialog.setLeftBtn("取消", new OnSimpleListener() {
//        @Override
//        public void onCallback() {
//
//        }
//    });
//    dialog.show();
//
// kotlin ----------------------------------------
//SimpleDialog().apply {
//    title = "我是标题"
//    message = "猜猜我是谁"
//    leftButton = "取消"
//    rightButton = "确定"
//    rightClickListener = OnSimpleListener {
//
//    }
//    leftClickListener = OnSimpleListener {
//
//    }
//}.show()
class SimpleDialog : BaseDialog<DialogSimpleBinding>() {
    @ColorRes
    var titleColor = R.color.text_1level
    var titleSize = 16f
    var title: CharSequence? = null

    @ColorRes
    var messageColor = R.color.text_3level
    var message: CharSequence? = null
    var messageSize = 14f
    var movementMethod: MovementMethod? = null

    @ColorRes
    var rightTextColor = 0
    var rightButton: CharSequence? = null
    var rightClickListener: OnSimpleListener? = null

    @ColorRes
    var leftTextColor = 0
    var leftButton: CharSequence? = null
    var leftClickListener: OnSimpleListener? = null

    override fun init() {
        if (TextUtils.isEmpty(title)) {
            binding.title.visibility = View.GONE
        } else {
            binding.title.text = title
            binding.title.textSize = titleSize
            binding.title.visibility = View.VISIBLE
            binding.title.setTextColorId(titleColor)
        }
        if (!TextUtils.isEmpty(message)) {
            binding.content.setTextColorId(messageColor)
            binding.content.text = message
            binding.content.post {
                if (binding.content.lineCount > 1) {
                    binding.content.gravity = Gravity.LEFT
                } else {
                    binding.content.gravity = Gravity.CENTER
                }
            }
            binding.content.textSize = messageSize
            binding.content.setTextColorId(R.color.text_2level)
            if (movementMethod != null) {
                binding.content.movementMethod = movementMethod
            } else {
                binding.content.movementMethod = ScrollingMovementMethod.getInstance()
            }
        }

        if (!TextUtils.isEmpty(rightButton)) {
            binding.rightButton.text = rightButton
            binding.rightButton.visibility = View.VISIBLE
            if (rightTextColor != 0) {
                binding.rightButton.setTextColor(getColor(rightTextColor))
            }
            binding.rightButton.setOnClickListener {
                cancel()
                rightClickListener?.onCallback()
            }
        } else {
            binding.rightButton.visibility = View.GONE
        }
        var isShowDivider = false
        if (!TextUtils.isEmpty(leftButton)) {
            isShowDivider = true
            binding.leftButton.text = leftButton
            binding.leftButton.visibility = View.VISIBLE
            if (leftTextColor != 0) {
                binding.leftButton.setTextColor(getColor(leftTextColor))
            }
            binding.leftButton.setOnClickListener {
                cancel()
                leftClickListener?.onCallback()
            }
        } else {
            binding.leftButton.visibility = View.GONE
        }
        if (isShowDivider) {
            binding.divider.visibility = View.VISIBLE
        } else {
            binding.divider.visibility = View.GONE
        }
    }

    /**
     * 开启单例模式
     */
    fun setSingle() {
        setId(uniqueDialogId())
    }

    /**
     * 根据调用类名和执行代码位置获取一个唯一的id
     */
    private fun uniqueDialogId(): Long {
        var dialogId = System.currentTimeMillis()
        val stackTraces = Throwable().stackTrace
        for (stackTrace in stackTraces) {
            val lineNumber = stackTrace.lineNumber
            if (lineNumber > 0) {
                val className = stackTrace.className
                try {
                    val clazz = Class.forName(className)
                    if (!filterStackTraceClass(clazz)) {
                        val hashCode = (stackTrace.fileName + ":" + lineNumber).hashCode()
                        dialogId = abs(hashCode.toDouble()).toLong()
                        log.i("对话框调用信息:当前class:$className,调用代码行数：$lineNumber")
                        break
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        log.i("生成对话框的id:$dialogId")
        return dialogId
    }

    private fun filterStackTraceClass(clazz: Class<*>): Boolean {
        return if (SimpleDialog::class.java == clazz) {
            true
        } else if (clazz.isInterface) {
            true
        } else {
            Modifier.isAbstract(clazz.modifiers)
        }
    }
}